### DISCUSSION TITLE
Change cache location (Node.js)

### DISCUSSION THREAD

**Discussion by janober** [pyodide/pyodide#3950] [General]
It looks like that when used with Node.js it caches the `.whl` files directly in the `node_modules/pyodide` folder. Is there any way to change that to a totally different folder or at least a subfolder?

Thanks for your help!

**Comment by ryanking13** [comment]
No, right now there is no way to modify the cache directory. For now, I would suggest you should download the full Pyodide archive that contains all packages, extract it to the directory that you want to store the package wheels, and [set `lockFileURL` argument](https://pyodide.org/en/stable/usage/api/js-api.html#globalThis.loadPyodide) when calling loadPyodide to the repodata.json file in that directory.

**Reply by janober** [reply]
Thanks a lot for your answer and for providing a possible workaround. But it would increase our Docker image a lot in size if we always ship it with all of the existing modules. So I assume we then have no choice than making the  `node_modules/pyodide` directory writable in the image, even if it is ugly, and it means modules get redownloaded after every restart.

Thanks for your help!

**Reply by ryanking13** [reply]
Or maybe you can try downloading the [pyodide-core](https://github.com/pyodide/pyodide/releases/tag/0.23.3) archive, extract it, and set the `indexURL` argument of `loadPyodide` to that directory. I think it will work, but I'm not very sure as there is a known bug in Node.js related to [modifying indexURL](https://github.com/pyodide/pyodide/issues/3550).

**Comment by netroy** [comment]
I tried adding a new config property called `installURL` (which defaults to `indexURL`), and then updated `downloadPackage` to use `API.config.installURL` instead of `API.config.indexURL`. 
This seems to work as expected. All new packages were downloaded to the path defined in `installURL`, while all the data packaged in the npm package was picked up from `indexURL`.

https://github.com/netroy/pyodide/commit/a8904fb7f7defe5f1cdcbf8d2f7cb10c1dd8e14d

I forked off `0.22.1`, as there seem to some breaking change in `0.23.x` for us.
But, if this looks acceptable to you, I can try to update our code base to work with `0.23.x`, and then send a PR with these changes.

**Reply by hoodmane** [reply]
Yes we'd accept a PR like that. Maybe call it `npmInstallDirectory` or `npmPackageCache` or something along those lines.

**Reply by netroy** [reply]
are you certain about the `npm` prefix? it might be confusing for the users as this folder will be used to store python packages.

**Reply by hoodmane** [reply]
No, I'm not certain. Open a PR with any name you like and we can bikeshed it as part of the review.

**Reply by netroy** [reply]
Created the PR https://github.com/pyodide/pyodide/pull/3967 

I'm not sure if `installURL` is a good name for what this config property does. so, I'd gladly update it to whatever name you suggest 🙏🏽

**Comment by rajsite** [comment]
Similarly related, I'm trying a node-like environment ([Deno Deploy with npm support](https://deno.com/blog/npm-on-deno-deploy)) where the `node_modules` directory is read-only (in addition to the [rest of the filesystem](https://deno.com/subhosting/docs/virtual-filesystem)).

Is there a way in the node context to use MEMFS for saving packages like the browser instead of trying to write to the filesystem?

Full example: 

```ts
import pyodideModule from "npm:pyodide/pyodide.js";
const { loadPyodide } = pyodideModule;
const pyodide = await loadPyodide();
await pyodide.loadPackage("numpy");
const result = await pyodide.runPythonAsync(`
import numpy as np
a = np.arange(15).reshape(3, 5)
a
`);
console.log("result:", result.toString());

Deno.serve(async (req: Request) => {
    return new Response("Hello World: " + result.toString());
});
```

Results in logs (oldest at bottom):
```
gcp-us-south1
PythonError: Traceback (most recent call last):
  File "/lib/python311.zip/_pyodide/_base.py", line 540, in eval_code_async
    await CodeRunner(
  File "/lib/python311.zip/_pyodide/_base.py", line 363, in run_async
    coroutine = eval(self.code, globals, locals)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<exec>", line 2, in <module>
ModuleNotFoundError: The module 'numpy' is included in the Pyodide distribution, but it is not installed.
You can install it by calling:
  await micropip.install("numpy") in Python, or
  await pyodide.loadPackage("numpy") in JavaScript
See https://pyodide.org/en/stable/usage/loading-packages.html for more details.

    at new_error (file:///node_modules/.deno/pyodide@0.23.4/node_modules/pyodide/pyodide.asm.js:9:14992)
    at <anonymous> (wasm://wasm/02239a6e:1:1387880)
    at <anonymous> (wasm://wasm/02239a6e:1:1416477)
    at _PyCFunctionWithKeywords_TrampolineCall (file:///node_modules/.deno/pyodide@0.23.4/node_modules/pyodide/pyodide.asm.js:9:128013)
    at <anonymous> (wasm://wasm/02239a6e:1:1701575)
    at <anonymous> (wasm://wasm/02239a6e:1:2647308)
    at <anonymous> (wasm://wasm/02239a6e:1:1967640)
    at <anonymous> (wasm://wasm/02239a6e:1:1702830)
    at <anonymous> (wasm://wasm/02239a6e:1:1703535)
    at <anonymous> (wasm://wasm/02239a6e:1:1703697)
    at <anonymous> (wasm://wasm/02239a6e:1:2509329)
    at <anonymous> (wasm://wasm/02239a6e:1:2533878)
    at <anonymous> (wasm://wasm/02239a6e:1:1703993)
    at <anonymous> (wasm://wasm/02239a6e:1:1703113)
    at <anonymous> (wasm://wasm/02239a6e:1:1414451)
    at Object.Module.callPyObjectKwargs (file:///node_modules/.deno/pyodide@0.23.4/node_modules/pyodide/pyodide.asm.js:9:75497)
    at Object.Module.callPyObject (file:///node_modules/.deno/pyodide@0.23.4/node_modules/pyodide/pyodide.asm.js:9:75942)
    at Timeout.wrapper [as _onTimeout] (file:///node_modules/.deno/pyodide@0.23.4/node_modules/pyodide/pyodide.asm.js:9:57697)
    at cb (ext:deno_node/internal/timers.mjs:63:31)
    at Object.action (ext:deno_web
9/11/2023, 3:51:38 PM
gcp-us-south1
Requires write access to '/node_modules/.deno/pyodide@0.23.4/node_modules/pyodide/', but the file system on Deno Deploy is read-only.
9/11/2023, 3:51:38 PM
gcp-us-south1
The following error occurred while loading numpy:
9/11/2023, 3:51:38 PM
gcp-us-south1
Failed to load numpy
9/11/2023, 3:51:38 PM
gcp-us-south1
Loading numpy
9/11/2023, 3:51:38 PM
```

**Reply by rth** [reply]
Do you know if Deno deploy provides any writable folders or is the whole file system read only? If yes, https://github.com/pyodide/pyodide/pull/3967 would address it.  If no, then indeed we need to downloaded files into the MEMFS. Currently it's not possible but I think we are interesting in making sure pyodide works in Deno deploy. Can you please open a separate issue about this please? Copy pasting your above comment as an issue would work. Thanks!
